apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Aecore.Util.StateTrees do
+  alias Aecore.Account.AccountStateTree
+  alias Aecore.Contract.CallStateTree
+  alias Aecore.Contract.ContractStateTree
+  alias Aecore.Channel.ChannelStateTree
+  alias Aecore.Naming.NamingStateTree
+
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  defmacro __using__(_) do
```

Comment:
The __using__ macro can have parameters. This can be used here to remove a lot of code and guarantee type safety. 
For instance by declaring the __using__ macro as:
```
defmacro __using__(tree_type, stored_type) when is_atom(tree_type) and is_module(stored_type) do
```
We can then write for instance:
```
PatriciaMerkleTree.new(tree_type)
```
instead of
```
PatriciaMerkleTree.new(StateTrees.tree_type(__MODULE__))
```
or more importantly:
```
{:ok, %stored_type{} = deserialized_value} = Serialization.rlp_decode_only(serialized_value, stored_type)
```
instead of:
```
{:ok, deserialized_value} = Serialization.rlp_decode_anything(serialized_value)
```
Then to use the new code you write:
```
use Aecore.Util.StateTrees, :non_existant_tree, RandomStoredType 
```

This approach together with unquoting/quoting code will make this module pretty small.
Please define an overwridable callback which returns an default value for an unknown key (the default implementation will return an error). 

As @cytadela8 said bellow make an callback to setup the deserialized value. For instance:
```
def setup_value(identifier, decoded_value) do
raise "Provide an custom implementation"
end

defoverridable setup_value:2
```